src/auth.rs

```diff
@@ -75,188 +78,270 @@ pub trait AuthorizationPolicy<T> {
-                        }
-                    }
-                    _ => Outcome::Failure((
-                        Status::BadRequest,
-                        anyhow!("Token action not matching endpoint"),
-                    )),
-                }
-            }
+    async fn from_request(req: &'l Request<'_>) -> Outcome<Self, Self::Error> {
+        let (config, relay) = match get_state(req) {
+            Ok(s) => s,
+            Err(e) => return internal_server_error(e),
+        };
+
+        // TODO: Support only zcaps.
+        let token = match SIWEDelegation::from_request(req).await {
+            Outcome::Success(t) => t,
+            Outcome::Failure(e) => return Outcome::Failure(e),
+            Outcome::Forward(_) => return unauthorized(anyhow!("no delegation found")),
+        };
+
+        let resource = match token
+            .delegation
+            .resources
+            .first()
+            .ok_or_else(|| anyhow!("delegation has empty resource list"))
+            .map(|uri| uri.as_str())
+            .map(ResourceId::from_str)
+        {
+            Err(e) => return bad_request(e),
+            Ok(Err(e)) => return bad_request(e),
+            Ok(Ok(resource)) => resource,
+        };
+        let orbit_id = resource.orbit();
+
+        //let token = match ZCAPDelegation::from_request(req).await {
+        //    Outcome::Success(t) => t,
+        //    Outcome::Failure(e) => return Outcome::Failure(e),
+        //    Outcome::Forward(_) => return unauthorized(anyhow!("no delegation found")),
+        //};
+
+        //let orbit_id = match token.delegation.property_set.capability_action.first() {
+        //    None => return unauthorized(anyhow!("delegation has empty capability action list")),
+        //    Some(resource) => resource.orbit(),
+        //};
+
+        let orbit = match load_orbit(orbit_id.get_cid(), &config, relay).await {
+            Ok(Some(o)) => o,
+            Ok(None) => return not_found(anyhow!("No Orbit found")),
+            Err(e) => return unauthorized(e),
+        };
+
+        let delegation = match (orbit.capabilities.root.clone(), token.delegation).try_into() {
+            Err(e) => return unauthorized(e),
+            Ok(d) => d,
+        };
+
+        //let delegation = match (token.delegation).try_into() {
+        //    Err(e) => return unauthorized(e),
+        //    Ok(d) => d,
+        //};
+
+        if let Err(e) = orbit
+            .capabilities
+            .transact(Updates::new([delegation], []))
+            .await
+        {
+            return internal_server_error(e);
         }
-    };
+        Outcome::Success(Self)
+    }
 }
 
-impl_fromreq!(PutAuthWrapper, "put");
-impl_fromreq!(GetAuthWrapper, "get");
-impl_fromreq!(MetadataAuthWrapper, "metadata");
-impl_fromreq!(DelAuthWrapper, "del");
-impl_fromreq!(ListAuthWrapper, "list");
+pub enum InvokeAuthWrapper {
+    Create(OrbitId),
```

Comment:
IMO this should be handled by `DelegateAuthWrapper`, but I understand this maintains the current behaviour so fair enough

Reply:
Yeah agreed. I had a go at converting it to a delegation, but I couldn't get it working. I'll put an issue in for this.